pinot-spi/src/main/java/org/apache/pinot/spi/env/CommonsConfigurationUtils.java

```diff
@@ -276,20 +334,65 @@ public static String recoverSpecialCharacterInPropertyValue(String value) {
     return value.replace("\0\0", ",");
   }
 
-  private static PropertiesConfiguration createPropertiesConfiguration(boolean setIOFactory,
-      boolean setDefaultDelimiter) {
+  /**
+   * creates the instance of the {@link org.apache.commons.configuration2.PropertiesConfiguration}
+   * with custom IO factory based on kind {@link org.apache.commons.configuration2.PropertiesConfiguration.IOFactory}
+   * and legacy list delimiter {@link org.apache.commons.configuration2.convert.LegacyListDelimiterHandler}
+   *
+   * @param setDefaultDelimiter sets the default list delimiter.
+   * @param ioFactoryKind IOFactory kind, can be null.
+   * @return PropertiesConfiguration
+   */
+  private static PropertiesConfiguration createPropertiesConfiguration(boolean setDefaultDelimiter,
+     @Nullable PropertyIOFactoryKind ioFactoryKind) {
     PropertiesConfiguration config = new PropertiesConfiguration();
 
-    // setting IO Reader Factory
-    if (setIOFactory) {
-      config.setIOFactory(new ConfigFilePropertyReaderFactory());
+    // setting IO Reader Factory of the configuration.
+    if (ioFactoryKind != null) {
+      config.setIOFactory(ioFactoryKind.getInstance());
     }
 
-    // setting DEFAULT_LIST_DELIMITER
+    // setting the DEFAULT_LIST_DELIMITER
     if (setDefaultDelimiter) {
       config.setListDelimiterHandler(new LegacyListDelimiterHandler(DEFAULT_LIST_DELIMITER));
     }
 
     return config;
   }
+
+  /**
+   * checks whether the configuration file first line is version header or not.
+   * @param file configuration file
+   * @return String
+   * @throws ConfigurationException exception.
+   */
+  private static String getConfigurationHeaderVersion(File file)
+      throws ConfigurationException {
+    String versionValue = DEFAULT_PROPERTIES_CONFIGURATION_HEADER_VERSION;
+    if (file.exists()) {
+      try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+        String fileFirstLine = reader.readLine();
+        // header version is written as a comment and start with '# '
+        String versionHeaderCommentPrefix = String.format("# %s", VERSION_HEADER_IDENTIFIER);
+        // check whether the file has the version header or not
+        if (StringUtils.startsWith(fileFirstLine, versionHeaderCommentPrefix)) {
+          String[] headerKeyValue = fileFirstLine.split(VERSIONED_CONFIG_SEPARATOR);
+          if (headerKeyValue.length == 2) {
+            versionValue = headerKeyValue[1];
+          }
+        }
+      } catch (IOException exception) {
+        throw new ConfigurationException(
+            String.format("Error occurred while reading configuration file %s ",
```

Comment:
nit: was the whitespace after `%s` left intentionally? otherwise can save `String.format()` and simply `"Error ..." + file.getName()` for short, but not a blocking comment any way.